### ISSUE TITLE
Guarantees of numerical approximation with N/evalf?

### ISSUE THREAD

**Issue by coproc** [sympy/sympy#19940]
I am trying to understand the guarantees of numerical approximation in sympy.
```python
>>> from sympy import *
>>> e1 = 1/(sqrt(2)-1) - sqrt(2)
>>> simplify(e1)
1
>>> f1 = N(e1); f1
1.00000000000000
>>> f1._prec
53
```
Does that mean that we have the guarantee `abs(e1 - 1) <= 0.5*10**(-53)`?

Second example:
```python
>>> e0 = e1 - 1
>>> simplify(e0)
0
>>> f0 = N(e0); f0
0.e-122
>>> f0._prec
1
```
Does that mean that we have the guarantee `abs(e0) <= 0.5*10**(-122)`? Where do the differences in accuracy/precision for `e1` and `e0` come from?

And what does the option `strict = True` add to the result?
```python
>>> N(e1, strict=True)
1.00000000000000
>>> N(e0, strict=True)
...
sympy.core.evalf.PrecisionExhausted: Failed to distinguish the expression: 

-sqrt(2) - 1 + 1/(-1 + sqrt(2))

from zero. Try simplifying the input, using chop=True, or providing a higher maxn for evalf
```
Why does the approximation of `e0` raise this exception, while the approximation of `e1` does not? It seems that any expression equal to `0` will result in this exception when setting `strict=True`. From the proposed solutions for avoiding `PrecisionExhausted` none seems to be satisfactory:
* simplifying `e0` works in this case, but it does not work for other expressions equal to `0`
* setting `chop=True` will destroy the precision guarantees?
* increasing `maxn` will only work for expressions not exactly equal to `0`

Can anyone shed some light on this?

**Comment by jksuom** [comment]
> Does that mean that we have the guarantee abs(e1 - 1) <= 0.5*10**(-53)?

I don't think that this could be guaranteed. `_prec` only gives the number of bits in the internal representation. It is does not give the number of *correct* bits in general because of rounding errors.

**Comment by coproc** [comment]
> > Does that mean that we have the guarantee abs(e1 - 1) <= 0.5*10**(-53)?
> 
> I don't think that this could be guaranteed. `_prec` only gives the number of bits in the internal representation. It is does not give the number of _correct_ bits in general because of rounding errors.

Ok, I understand. But do we have the guarantee that `abs(e1 - 1) <= 0.5*10**(-15)`? (when N/evalf was called with `n=15`)

**Comment by oscarbenjamin** [comment]
I think that `_prec` is supposed to indicate the number of correct digits. It isn't the same as the number of digits in the internal mpmath representation: the `_prec` is a part of SymPy's `Float` rather than mpmath's `mpf`:
```julia
In [42]: Float(3, 5)                                                                                                                           
Out[42]: 3.0000

In [43]: Float(3, 5)._prec                                                                                                                     
Out[43]: 20

In [44]: Float(3, 5)._mpf_                                                                                                                     
Out[44]: (0, 3, 0, 2)

In [45]: Float(3, 10)                                                                                                                          
Out[45]: 3.000000000

In [64]: Float(3, 10)._prec                                                                                                                    
Out[64]: 37

In [65]: Float(3, 10)._mpf_                                                                                                                    
Out[65]: (0, 3, 0, 2)
```
The intention is that if you ask for `N` digits of precision then evalf will adaptively ask mpmath to evalute the result using higher precision so that the result ends up having `N` correct digits. Sometimes that isn't possible because `_evalf` has a maximum internal precision setting and in those cases the returned result will have `_prec` less than what was requested. This is essentially guaranteed to happen when evaluating an expression that is identically zero.

Note that the precision argument to evalf is decimal digits but the `_prec` property of `Float` is binary digits:
```julia
In [58]: pi.n(10)                                                                                                                              
Out[58]: 3.141592654

In [59]: pi.n(10)._prec                                                                                                                        
Out[59]: 37

In [63]: N(37 / log(10, 2))                                                                                                                    
Out[63]: 11.1381098395673
```

SymPy's Float and evalf do not dinstinguish between "accuracy" and "precision" in the way that Mathematica does:
https://reference.wolfram.com/language/howto/ControlThePrecisionAndAccuracyOfNumericalResults.html

**Comment by jksuom** [comment]
> I think that _prec is supposed to indicate the number of correct digits.

That may have been originally intended but it is not guaranteed by the current implementation. The arithmetic operations on Floats make use of this private method:
https://github.com/sympy/sympy/blob/8544f01468187b9b27d4fdb4671ce672d03622f9/sympy/core/numbers.py#L668-L670
Note that the `max` of the two precisions is taken.

**Comment by oscarbenjamin** [comment]
I'm not sure what that does but if it violates the guarantee then I would rather consider that a bug that should be fixed. The intention should be to guarantee the meaning of `_prec` in some documented way.

**Comment by coproc** [comment]
> I think that `_prec` is supposed to indicate the number of correct digits. 
> [...]
> Note that the precision argument to evalf is decimal digits but the `_prec` property of `Float` is binary digits:
> [...]

That would mean that `f = N(e, n=p)` guarantees (should guarantee) something like `abs((e - f)/e) < 2**(-f._prec)`. And for the printed decimal representation `fs` of `f` (with `p` decimal places) we get an additional maximal relative rounding error of `5*10**(-p) `, altogether: `abs((e - S(fs))/e) < 2**(-f._prec) + 5*10**(-p)`. Is this correct?

**Comment by oscarbenjamin** [comment]
It should be something roughly like that. Note that ideally the `_prec` should be slightly greater than `p * log(10, 2)`. If it is then the printed form should give `p` decimal digits. Otherwise the printed form will give `p2` decimal digits where I think `p2` is slightly less than `_prec / log(10, 2)`.
